internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -100,159 +102,204 @@ func (r *Reconciler) getNewMachineSet(ctx context.Context, d *clusterv1.MachineD
+	for k, v := range labels {
+		desiredMS.Labels[k] = v
 	}
-
 	// Enforce that the MachineDeploymentNameLabel label is set
 	// Note: the MachineDeploymentNameLabel is added by the default webhook to MachineDeployment.spec.template.labels if spec.selector is empty.
-	newMS.Labels[clusterv1.MachineDeploymentNameLabel] = d.Name
+	desiredMS.Labels[clusterv1.MachineDeploymentNameLabel] = deployment.Name
+
+	// Set annotations
+	desiredMS.Annotations = mdutil.CalculateMachineSetAnnotations(deployment, newRevision, existingMS)
 
-	if d.Spec.Strategy.RollingUpdate.DeletePolicy != nil {
-		newMS.Spec.DeletePolicy = *d.Spec.Strategy.RollingUpdate.DeletePolicy
+	// Set MinReadySeconds
+	if deployment.Spec.MinReadySeconds != nil {
+		desiredMS.Spec.MinReadySeconds = *deployment.Spec.MinReadySeconds
 	}
 
-	// Add foregroundDeletion finalizer to MachineSet if the MachineDeployment has it
-	finalizerSet := sets.Set[string]{}.Insert(d.Finalizers...)
-	if finalizerSet.Has(metav1.FinalizerDeleteDependents) {
-		controllerutil.AddFinalizer(&newMS, metav1.FinalizerDeleteDependents)
+	// Set DeletePolicy
+	if deployment.Spec.Strategy.RollingUpdate.DeletePolicy != nil {
+		desiredMS.Spec.DeletePolicy = *deployment.Spec.Strategy.RollingUpdate.DeletePolicy
 	}
 
-	allMSs := append(oldMSs, &newMS)
-	newReplicasCount, err := mdutil.NewMSNewReplicas(d, allMSs, &newMS)
-	if err != nil {
-		return nil, err
+	// Set Node timeout values
+	desiredMS.Spec.Template.Spec.NodeDrainTimeout = deployment.Spec.Template.Spec.NodeDrainTimeout
+	desiredMS.Spec.Template.Spec.NodeDeletionTimeout = deployment.Spec.Template.Spec.NodeDeletionTimeout
+	desiredMS.Spec.Template.Spec.NodeVolumeDetachTimeout = deployment.Spec.Template.Spec.NodeVolumeDetachTimeout
+
+	// Set controller ref.
+	controllerRef := *metav1.NewControllerRef(deployment, machineDeploymentKind)
+	desiredMS.OwnerReferences = util.EnsureOwnerRef(desiredMS.OwnerReferences, controllerRef)
+
+	// Add foregroundDeletion finalizer to MachineSet if the MachineDeployment has it
+	if sets.New[string](deployment.Finalizers...).Has(metav1.FinalizerDeleteDependents) {
+		controllerutil.AddFinalizer(desiredMS, metav1.FinalizerDeleteDependents)
 	}
 
-	*(newMS.Spec.Replicas) = newReplicasCount
+	if existingMS == nil {
+		// Creating a new MachineSet
+		desiredMS.SetName(names.SimpleNameGenerator.GenerateName(deployment.Name + "-"))
 
-	// Set new machine set's annotation
-	mdutil.SetNewMachineSetAnnotations(d, &newMS, newRevision, false, log)
-	// Create the new MachineSet. If it already exists, then we need to check for possible
-	// hash collisions. If there is any other error, we need to report it in the status of
-	// the Deployment.
-	alreadyExists := false
-	err = r.Client.Create(ctx, &newMS)
-	createdMS := &newMS
-	switch {
-	// We may end up hitting this due to a slow cache or a fast resync of the Deployment.
-	case apierrors.IsAlreadyExists(err):
-		alreadyExists = true
+		machineSetUniqueIdentifier := apirand.String(10)
```

Comment:
Q: Can we re-use the same number we are using as a suffix for the MS name?

Reply:
What if the deployment.Name is very close to max-length. Is there a way to identify the suffix then?

Reply:
Ah I see. Essentially implement it ourselves instead of using SimpleNameGenerator, makes sense